File: ios/Runner/Native/BaseNativeHandler.swift
import Flutter
import UIKit

protocol BaseNativeHandler {
    var channelName: String { get }
    var eventChannels: [String: String] { get }
    
    func initialize()
    func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult)
    func getStreamHandler(eventName: String) -> (any FlutterStreamHandler & NSObjectProtocol)?
    func onDestroy()
}

class AbstractNativeHandler: NSObject, BaseNativeHandler {
    let flutterEngine: FlutterEngine
    let context: AppDelegate
    
    var methodChannel: FlutterMethodChannel!
    var eventChannelMap: [String: FlutterEventChannel] = [:]
    
    // Abstract properties - must be overridden
    var channelName: String {
        fatalError("Must override channelName in \(type(of: self))")
    }
    
    var eventChannels: [String: String] {
        return [:]
    }
    
    init(flutterEngine: FlutterEngine, context: AppDelegate) {
        self.flutterEngine = flutterEngine
        self.context = context
        super.init()
    }
    
    func initialize() {
        setupMethodChannel()
        setupEventChannels()
        print("🚀 Initialized handler: \(channelName)")
    }
    
    private func setupMethodChannel() {
        methodChannel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: flutterEngine.binaryMessenger
        )
        
        methodChannel.setMethodCallHandler { [weak self] call, result in
            guard let self = self else { return }
            print("📞 Method call: \(call.method) on \(self.channelName)")
            self.handleMethodCall(call, result: result)
        }
    }
    
    private func setupEventChannels() {
        eventChannels.forEach { (name, channel) in
            let eventChannel = FlutterEventChannel(
                name: channel,
                binaryMessenger: flutterEngine.binaryMessenger
            )
            eventChannelMap[name] = eventChannel
            
            // Fixed: Properly handle FlutterStreamHandler
            if let streamHandler = getStreamHandler(eventName: name) {
                eventChannel.setStreamHandler(streamHandler)
                print("🔄 Setup event channel: \(channel)")
            }
        }
    }
    
    // Abstract methods - must be overridden
    func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        fatalError("Must override handleMethodCall in \(type(of: self))")
    }
    
    func getStreamHandler(eventName: String) -> (any FlutterStreamHandler & NSObjectProtocol)? {
        return nil
    }
    
    func onDestroy() {
        eventChannelMap.removeAll()
        print("💥 Destroyed handler: \(channelName)")
    }
}

File: android/app/src/main/kotlin/com/example/news_app/native/BaseNativeHandler.kt
package com.nytimes.news_app.native

import android.content.Context
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

abstract class BaseNativeHandler(
    protected val flutterEngine: FlutterEngine,
    protected val context: Context
) {
    abstract val channelName: String
    abstract val eventChannels: Map<String, String>

    protected lateinit var methodChannel: MethodChannel
    protected val eventChannelMap = mutableMapOf<String, EventChannel>()

    open fun initialize() {
        setupMethodChannel()
        setupEventChannels()
        println("✅ Android: Initialized handler: $channelName")
    }

    private fun setupMethodChannel() {
        methodChannel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, channelName)
        methodChannel.setMethodCallHandler { call, result ->
            println("📞 Android: Method call: ${call.method} on $channelName")
            handleMethodCall(call, result)
        }
    }

    private fun setupEventChannels() {
        eventChannels.forEach { (name, channel) ->
            val eventChannel = EventChannel(flutterEngine.dartExecutor.binaryMessenger, channel)
            eventChannelMap[name] = eventChannel
            eventChannel.setStreamHandler(getStreamHandler(name))
            println("🔄 Android: Setup event channel: $channel")
        }
    }

    abstract fun handleMethodCall(call: MethodCall, result: Result)
    open fun getStreamHandler(eventName: String): EventChannel.StreamHandler? = null
    open fun onDestroy() {
        eventChannelMap.clear()
        println("💥 Android: Destroyed handler: $channelName")
    }
}